src/script/sign.h

```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+
+        // Write script sig
+        if (!final_script_sig.empty()) {
+            SerializeToVector(s, PSBT_IN_SCRIPTSIG);
+            s << final_script_sig;
+        }
+        // write script witness
+        if (!final_script_witness.IsNull()) {
+            SerializeToVector(s, PSBT_IN_SCRIPTWITNESS);
+            SerializeToVector(s, final_script_witness.stack);
+        }
+
+        // Write unknown things
+        for (auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        s << PSBT_SEPARATOR;
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        while(!s.empty()) {
+            // read size of key
+            uint64_t key_len = ReadCompactSize(s);
+
+            // the key length is 0 if that was actually a separator byte
+            // This is a special case for key lengths 0 as those are not allowed (except for separator)
+            if (key_len == 0) {
+                return;
+            }
+
+            // Read key
+            std::vector<unsigned char> key;
+            key.resize(key_len);
+            s >> MakeSpan(key);
+
+            // First byte of key is the type
+            unsigned char type = key[0];
+
+            // Do stuff based on type
+            switch(type) {
+                case PSBT_IN_NON_WITNESS_UTXO:
+                    if (non_witness_utxo) {
+                        throw std::ios_base::failure("Duplicate Key, input non-witness utxo already provided");
+                    }
+                    UnserializeFromVector(s, non_witness_utxo);
+                    break;
+                case PSBT_IN_WITNESS_UTXO:
+                    if (!witness_utxo.IsNull()) {
+                        throw std::ios_base::failure("Duplicate Key, input witness utxo already provided");
+                    }
+                    UnserializeFromVector(s, witness_utxo);
+                    break;
+                case PSBT_IN_PARTIAL_SIG:
+                {
+                    // Make sure that the key is the size of pubkey + 1
+                    if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
+                        throw std::ios_base::failure("Size of key was not the expected size for the type partial signature pubkey");
+                    }
+                    // Read in the pubkey from key
+                    CPubKey pubkey(key.begin() + 1, key.end());
+                    if (!pubkey.IsFullyValid()) {
+                       throw std::ios_base::failure("Invalid pubkey");
+                    }
+                    if (partial_sigs.count(pubkey.GetID()) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, input partial signature for pubkey already provided");
+                    }
+
+                    // Read in the signature from value
+                    uint64_t value_len = ReadCompactSize(s);
```

Comment:
I think this can be written as just `std::vector<unsigned char> sig; s >> sig`.

Reply:
Done